src/scp/QuorumIntersectionChecker.cpp

```diff
@@ -0,0 +1,1106 @@
+            {
+                scc(j);
+                v.mLowLink = std::min(v.mLowLink, w.mLowLink);
+            }
+            else if (w.mOnStack)
+            {
+                v.mLowLink = std::min(v.mLowLink, w.mIndex);
+            }
+        }
+
+        if (v.mLowLink == v.mIndex) {
+            BitSet newScc;
+            newScc.set(i);
+            size_t j = -1;
+            do
+            {
+                j = mStack.back();
+                newScc.set(j);
+                mStack.pop_back();
+                mNodes.at(j).mOnStack = false;
+            }
+            while (j != i);
+            mSCCs.push_back(newScc);
+        }
+    }
+};
+
+class QuorumIntersectionCheckerImpl : public QuorumIntersectionChecker
+{
+
+    Config const& mCfg;
+
+    struct Stats {
+        size_t mTotalNodes = {0};
+        size_t mNumSCCs = {0};
+        size_t mMaxSCC = {0};
+        size_t mCallsStarted = {0};
+        size_t mFirstRecursionsTaken = {0};
+        size_t mSecondRecursionsTaken = {0};
+        size_t mMinQuorumsSeen = {0};
+        size_t mTerminations = {0};
+        size_t mEarlyExit1s = {0};
+        size_t mEarlyExit21s = {0};
+        size_t mEarlyExit22s = {0};
+        size_t mEarlyExit31s = {0};
+        size_t mEarlyExit32s = {0};
+        void log() const;
+    };
+
+    // We use our own stats and a local cached flag to control tracing because
+    // using the global metrics and log-partition lookups at a fine grain
+    // actually becomes problematic CPU-wise.
+    mutable Stats mStats;
+    bool mLogTrace;
+
+    // These are the key state of the checker: the mapping from node public keys
+    // to graph node numbers, and the graph of QBitSets itself.
+    std::vector<PublicKey> mBitNumPubKeys;
+    std::unordered_map<PublicKey, size_t> mPubKeyBitNums;
+    QGraph mGraph;
+
+    // This just calculates SCCs and stores the maximal one, which we use for
+    // the remainder of the search.
+    TarjanSCCCalculator mTSC;
+    BitSet mMaxSCC;
+
+    QBitSet convertSCPQuorumSet(SCPQuorumSet const& sqs);
+    void buildGraph(QuorumTracker::QuorumMap const& qmap);
+    void buildSCCs();
+
+    bool containsQuorumSlice(BitSet const& bs, QBitSet const& qbs) const;
+    bool containsQuorumSliceForNode(BitSet const& bs, size_t node) const;
+    BitSet contractToMaximalQuorum(BitSet nodes) const;
+    bool isAQuorum(BitSet const& nodes) const;
+    bool isMinimalQuorum(BitSet const& nodes) const;
+    bool hasDisjointQuorum(BitSet const& nodes) const;
+    void logLoudErrorAboutDisjointQuorums(BitSet const& nodes,
+                                          BitSet const& disj) const;
+    std::string nodeName(size_t node) const;
+
+    friend class MinQuorumEnumerator;
+
+public:
+    QuorumIntersectionCheckerImpl(QuorumTracker::QuorumMap const& qmap,
+                                  Config const& cfg)
+        : mCfg(cfg),
+          mLogTrace(Logging::logTrace("SCP")),
+          mTSC(mGraph)
```

Comment:
isn't mGraph empty here?

Reply:
Yes, but we pass it by reference; it's populated by the time mTSC reads it.